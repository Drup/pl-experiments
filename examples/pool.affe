type ('a : 'k) option : 'k = None of unit | Some of 'a

type ('a : 'k) queue : 'k 

val qcreate : \ ('a : _) . unit -> 'a queue
val qpush : \ 'k2 ('a : 'k) . 'a -> &('k2, 'a queue) -{'k}> unit
val qpop : \ 'k2 ('a : _) . &('k2, 'a queue) -> 'a option
val qiter : \ ('a :_) . ('a -> unit) -> 'a queue -> unit

type ('a: 'k) pool : 'k = Pool of ((unit -> 'a) * 'a queue)
let create f = Pool (f, qcreate())

let consume f c = match c with {
  | Pool (_cre,q) -> qiter f q
  }

let use p f = 
  match p with { Pool (cr, q) ->
  let o = 
  match qpop &q with {
    | Some x -> x
    | None () -> cr ()
  } in
  let r = f &!o in
  let () = qpush o &q in
  r
  }
