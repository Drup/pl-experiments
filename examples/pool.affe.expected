type option = ∀^k. ^k -> ^k
constructor None : ∀'a. unit -> 'a option
constructor Some : ∀'a. 'a -> 'a option
type queue = ∀^k. ^k -> ^k
qcreate : ∀^k, 'a. ('a : ^k) => unit -> 'a queue
qpush : ∀^k, ^k2, 'a. ('a : ^k) => 'a -> &(^k2,'a queue) -{^k}> unit
qpop : ∀^k, ^k2, 'a. ('a : ^k) => &(^k2,'a queue) -> 'a option
qiter : ∀^k, 'a. ('a : ^k) => ('a -> unit) -> 'a queue -> unit
type pool = ^_k -> ^_k
constructor Pool : ∀'a. (unit -> 'a) * 'a queue -> 'a pool
create : ∀^k, 'a. ('a : ^k) => (unit -> 'a) -> 'a pool

consume : ∀^k, 'a. ('a : ^k) => ('a -> unit) -> 'a pool -> unit

use :
  ∀^k, ^l, ^m, 'a, 'a1. ('a1 : ^m) & ('a : ^l) & (^l < lin₁) =>
  'a1 pool -> (&!(aff₄,'a1) -{^k}> 'a) -{^m}> 'a

